### ISSUE TITLE
Forcing all resources to be serializable seems overly restrictive

### ISSUE THREAD

**Issue by danheuck** [gbj/leptos#34]
It seems like resources are currently forced to be serializable to support server side rendering (and potentially resumability in the future). This makes sense, but it excludes some use cases that would otherwise make sense as resources. For example, I was experimenting with accessing a webcam in leptos (using nokhwa). This is an async operation and, reading the docs, it seemed like a resource was a perfect fit. However, a webcam handle isn't serializable so I ended up with this slightly more awkward implementation:

``` rust
use leptos::*;
use nokhwa::{js_camera::JSCameraConstraintsBuilder, JSCamera};
use std::{cell::RefCell, rc::Rc};


async fn get_camera(set_camera: WriteSignal<Option<Rc<RefCell<JSCamera>>>>) {
    let constraints = JSCameraConstraintsBuilder::new().build();
    let camera = JSCamera::new(constraints).await;

    match camera {
        Ok(camera) => {
            log::info!("Got camera");
            set_camera(Some(Rc::new(RefCell::new(camera))));
        }
        Err(e) => {
            log::error!("Failed to get camera: {}", e);
        }
    }
}

#[component]
pub fn UserWebcam(cx: scope) -> Element {
    let (camera, set_camera) = create_signal::<Option<Rc<RefCell<JSCamera>>>>(cx, None);
    spawn_local(get_camera(set_camera));

    create_effect(cx, move |_| {
        if let Some(camera) = camera.get() {
            if let Err(e) = (*camera).borrow_mut().attach("camera_out", false) {
                log::error!("Failed to attach camera: {}", e);
            } else {
                log::info!("Attached camera");
            }
        }
    });

    view! {cx,
        <video id="camera_out" class="user_webcam"/>
    }
}
```
It works, but it's not ideal and it can't integrate with functionality like suspense contexts (unless I'm missing something).

It feels like the most natural solution would be to split resources into serializable and non-serializable types. Serializable resources would keep the current behavior but non-serializable ones would be handled fully on the client side regardless of rendering configuration. Happy to take a stab at implementing this if you're open to a PR, but figured I'd open an issue first to at least make sure this functionality doesn't already exist somewhere.

**Comment by gbj** [comment]
1. This is really cool.
2. Definitely go ahead and make a PR. You should be able to base things on the basic [Resource]: maybe `create_client_resource()` or something, and including a simple `bool` flag that the resource can check re: serialization. Basically the main thing I can think of it that this should be *excluded* from the set of "pending resources" sent from the server during SSR... If it's not in the "pending resources" then on hydration it will just be run on the client, and in CSR it will just be run on the client.

Let me know if you need any help navigating the code or if it's too much and you want me to take a stab at it. Apologies for the lack of docs on Resources so far.

**Comment by gbj** [comment]
A few notes in case they're useful

Probably put a bool client flag in here
https://github.com/gbj/leptos/blob/794ad5b2b2f39ca43f39c809c8b2dd3fd962171e/leptos_reactive/src/resource.rs#L301-L318

We need to figure out what to do here and below
https://github.com/gbj/leptos/blob/794ad5b2b2f39ca43f39c809c8b2dd3fd962171e/leptos_reactive/src/resource.rs#L144

Maybe this should return `Option<...>`?
https://github.com/gbj/leptos/blob/794ad5b2b2f39ca43f39c809c8b2dd3fd962171e/leptos_reactive/src/resource.rs#L408-L421

Filter out client resources here
https://github.com/gbj/leptos/blob/794ad5b2b2f39ca43f39c809c8b2dd3fd962171e/leptos_reactive/src/runtime.rs#L220-L232

**Comment by danheuck** [comment]
Just put up a PR for this (#35), let me know what you think.

I do agree that the ` Option<Rc<RefCell>>` is far from ideal. There might be a better way to represent it without modifying the library code at all, there's still a lot of Rust for me to learn. Dropping the clones in the Signal API could be nice, but I'm not sure whether copying/taking ownership are any better, I'll experiment with it a bit if I have time.

**Comment by gbj** [comment]
Closed by #35
